### ISSUE TITLE
memoize caching the python object, not the JSON serialization

### ISSUE THREAD

**Issue by sdementen** [thedirtyfew/dash-extensions#16]
I am very interested by the memoize capacity yet by going through the code, I see it is not caching the JSONResponse but the object before its serialization.
As the JSON serialiaztion is quite expensive in my case (hourly data on a year with multiple traces => lot of datetime JSON serialisation), it is still very slow.
Would it be feasible to cache the final response (JSON) instead of the intermediate `Figure` object ?

**Comment by emilhe** [comment]
If the JSON serialization is in fact expensive, i guess it would make sense to cache the JSON string rather than the object. That being said, i haven't encountered any real life examples where the JSON serialization itself was a performance bottleneck. Are you sure that's the case? How have you measured it?

**Comment by sdementen** [comment]
I have measured the time for the callback to return (with a timing
decorator) and the time for the request (with a
before_request/after_request timing).
When cached, the first time is close to zero but the second is around
4seconds, time that I also see in the browser dev tool.

**Comment by emilhe** [comment]
To check the json serializaton time, i would suggest measuring it directly in the callback, i.e. something like

```
import json
import datetime

obj = {}
tic = datetime.datetime.now()
json_string = json.dumps(obj)
toc = datetime.datetime.now()
print(f"TIME TO SERIALIZE = {(toc-tic).total_seconds()}")
```

where obj is the figure in question. As you measure it now, i guess that the data transfer time is included?

**Comment by sdementen** [comment]
The timing probably includes the transfer time but it is on localhost. Do
you think it can be significant?
I will time tomorrow the json serialization as you suggest.

**Comment by emilhe** [comment]
Yes, in most cases (as per my experience), it takes much longer than the serialization itself.

**Comment by sdementen** [comment]
I have done some measurements by adding time logs around the figure
generation itself, around the whole callback and around the json.dumps in
Dash callback wrapper:
--------    'figure generation' started
********    'figure generation' ran in 3.912 seconds
func:'update_output_div_grid_graph' took: 4.3776 sec
--------    'dumping response' started
********    'dumping response' ran in 2.704 seconds
duration of request /_dash-update-component = 7.1836 s

So, my callback takes 4.3s (of which 3.9s generating the plotly figure).
Serialiazing to JSON takes 2.7s
Total request is 7.2s
Transfer time = 7.2-2.7-4.3=0.2s       (447kb compressed, 2.6M uncompressed)

without memoization => 7.2s
With the current memoize behavior, I save on the 4.3s of my callback but
still have the json & the transfer => 2.7+0.2 = 2.9s
If the memoization could be done after the JSON serialisation => 0.2s

**Comment by emilhe** [comment]
Wow, that is slow for the serialization! To do the serialization before caching, I guess you could just return the serialized figure, i.e call json.dumps on the figure object.

**Comment by sdementen** [comment]
I have tried to return the jsonified figure but then Dash will see a string
and jsonify again the string which the react component do not like

**Comment by sdementen** [comment]
The slowness of the serialization is maybe due to the fact that there are 9
traces, built from dataframes at an hourly level on a full year (index=8760
datetimes that are serialised to json 9 times)

**Comment by sdementen** [comment]
ok, I have noticed two points that takes time:
- serializing the dataframe datetimeindex ==> I convert the index to string in the original dataframe before creating the figure 
- plotly json conversion (vs the standard one) ==> it takes for instance 0.28s instead of 0.08s to dumps via plotly instead of the std json package ==> I do a `fig = json.loads(json.dumps(fig, cls=plotly.utils.PlotlyJSONEncoder))` within the callback (a bit like what you suggested) and I save 0.2s
With these two hacks, it is now much more responsive when memoize.
Maybe the last hack could integrate dash-extensions as an option for the memoize argument with True (current behavior for memoization), False (no memoization), "prejson" (prejsonify before returning)

**Comment by emilhe** [comment]
Haha, that’s a nice hack 👍

**Comment by sdementen** [comment]
for the record, I "integrated" it in dash_extensions by adding after https://github.com/thedirtyfew/dash-extensions/blob/master/dash_extensions/enrich.py#L393 the following;
```
            if memoize == "prejson":
                data = json.loads(json.dumps(data, cls=plotly.utils.PlotlyJSONEncoder))
```
Now, if I use @app.callback(..., memoize="prejson"), it will memoize after applying the JSON dump/load hack.

Do you think it can be integrated in the package (to avoid me "monkey patching" your package in my project) ?

**Comment by emilhe** [comment]
Jep. Could you provide a small self contained example of the usage?

**Comment by sdementen** [comment]
Here is an example where you can "refresh" a graph with the three memoize modes (True, False, "prejson"). You need to have patched your code as explained hereabove to have the "prejson" working.

What is interesting to notice (and for this you need to add some `with Timer("..."):` lines in your code) is that, once the figure has been converted to pure JSON (with the plotly dumps followed by the loads), it is also much faster to pickle (the timing of line https://github.com/thedirtyfew/dash-extensions/blob/master/dash_extensions/enrich.py#L387 is oing from 0.62s to 0.01s) even through if the pickled file is larger (probably because there is no more sharing of common objects).

So I am even wondering if the "prejson" mode should not be the default mode.

```
import functools
import time
from threading import get_ident

import dash_bootstrap_components as dbc
import dash_core_components as dcc
import dash_html_components as html
import numpy
import pandas
from dash_extensions.enrich import Dash, Output, Trigger
from flask import g, request

from dash_utils import px_area


class Timer:
    def __init__(self, desc_or_function="anonymous code block"):
        if isinstance(desc_or_function, str):
            # called as context
            self.description = desc_or_function
        else:
            # called as decorator
            self.description = desc_or_function.__name__
            self.obj = desc_or_function
            functools.update_wrapper(self, desc_or_function)
        self.log = print

    def __enter__(self):
        self.start = time.perf_counter()
        return self

    def __exit__(self, *args):
        self.end = time.perf_counter()
        self.interval = self.end - self.start
        self.log(f"Thread: '{get_ident()}'   '{self.description}' ran in {round(self.interval, 3)} seconds")

    def __call__(self, *args, **kwargs):
        # decorator call
        with Timer(self.description):
            return self.obj(*args, **kwargs)


df = pandas.DataFrame(
    index=pandas.date_range("2020-01", "2021-01", freq="H", closed="left", tz="CET"),
    columns=list("ABCDEFGHIJK"),
    data=pandas.np.nan,
)
df[:] = numpy.random.random(df.values.shape)

app = Dash(__name__)

memoize_modes = [True, "prejson", False, ]

app.layout = dbc.Container(
    [
        dbc.Row(
            [
                dbc.Col(
                    [
                        html.Button(id=f"trigger-{mode}", children=f"Refresh {mode}"),
                        dcc.Graph(id=f"my-graph-{mode}"),
                    ],
                    className="col-4",
                )
                for mode in memoize_modes
            ]
        )
    ]
)
for mode in memoize_modes:
    @app.callback(
        Output(component_id=f"my-graph-{mode}", component_property="figure"),
        Trigger(component_id=f"trigger-{mode}", component_property="n_clicks"),
        memoize=mode,
    )
    @Timer
    def update_graph():
        with Timer("figure generation"):
            fig = px_area(df)
        return fig


@app.server.before_request
def before_request():
    if request.path != "/_reload-hash":
        g.timer = Timer(request.path).__enter__()


@app.server.teardown_request
def teardown_request(exception=None):
    if request.path != "/_reload-hash":
        g.timer.__exit__()


if __name__ == "__main__":
    app.run_server(debug=True)

```

**Comment by emilhe** [comment]
Could you add the `px_area` also? I just tried with simple scatter plots, but they render pretty fast

**Comment by emilhe** [comment]
I just did a few test, and the speed up is indeed remarkable! It just made a build with the new feature,

https://pypi.org/project/dash-extensions/0.0.32rc6/

In this build i am using the syntax that you can pass _function handles_ via the `memoize` keyword. Here is a small example,

```
import itertools
import dash_core_components as dcc
import dash_html_components as html
import plotly.graph_objs as go
import pandas as pd
import numpy as np

from dash_extensions.enrich import Dash, Output, Trigger, plotly_jsonify


def make_figure():
    fig = go.Figure()
    for col in df.columns:
        fig.add_trace(go.Scatter(x=df.index, y=df[col]))
    return fig


def make_layout(tag):
    return [dcc.Graph(id=f"graph_{tag}"), html.Button(f"Update [{tag}]", id=f"btn_{tag}")]


tags = ["default", "memoize", "prejson"]
memoize = dict(default=None, memoize=True, prejson=plotly_jsonify)
# Create dummy data.
df = pd.DataFrame(index=pd.date_range("2020-01", "2021-01", freq="H"), columns=list("ABCDEFGHIJK"), data=pd.np.nan)
df[:] = np.random.random(df.values.shape)
# Create example app.
app = Dash(prevent_initial_callbacks=True)
app.layout = html.Div(list(itertools.chain.from_iterable([make_layout(tag) for tag in tags])))
# Attach callbacks.
for t in tags:
    app.callback(Output(f"graph_{t}", "figure"), Trigger(f"btn_{t}", "n_clicks"), memoize=memoize[t])(make_figure)

if __name__ == '__main__':
    app.run_server(port=8877, debug=True)
```

Let me know if this syntax is OK for your use case. I am still open to suggestions :)

**Comment by sdementen** [comment]
I do not see the code on the repo yet, could you push it on a branch ?
For the syntax, I understand that you want to be able to provide:
- None => as False before
- True => as True before (memoize the output of the callback without any postprocessing)
- a function => as True but where the output of the callback is passed to the function to get back the object to memoize
Is that correct ?
If so, the function ability may appear to be bring some ability to be flexible but I wonder if in this case it is not "premature generalisation". The user will never guess by himself that it can pass plotly_jsonify to get an improvement in speed.
I would rather have as possible values for memoize:
- True/"jsonify" => optimised behavior with the plotly_jsonify
- False/None/"none" => no memoization (=default)
- "raw" => memoize without the plotly_jsonify
(ie keep strings or plain bool/None as parameters with a good docstring, like in the pandas lib)

But I may have misunderstood the syntax

**Comment by emilhe** [comment]
I pushed it to a new branch yes,

https://github.com/thedirtyfew/dash-extensions/tree/prejson

The concept is like you note for [None/false, True, and function]. Hence to get the behavior you need, you would should pass `memoize=plotly_jsonify`. Setting `memoize="raw"` would also work, since `"raw"` would evaluate to `True` in a boolean conversion. I figured this syntax would accommodate if anyone needs a different pre processing (prior to caching), whatever that might be.

**Comment by sdementen** [comment]
I would rather adapt the syntax:
- True/"jsonify" => optimised behavior with the plotly_jsonify
- False/None/"none" => no memoization (=default)
- "raw" => memoize without the plotly_jsonify

And so:
- not add possibility for generic functions (not useful today, more complex for beginners, better to get to know new use cases that would need other preprocessing function via a github issue ;-) ... Otherwise you will not be aware of people using other inventive preprocessing functions like me)
- when True (the basic use of memoize), use the plotly_jsonify preprocessing, i.e. automatically enable the optimization
- consider "raw" as memoize but without any preprocessing (i.e. today behavior when giving True)

But it is also a matter of taste, so feel free to choose the one that makes the more sense

**Comment by sdementen** [comment]
hello @emilhe 
any news regarding this MR? tell me if I can be of any help!

**Comment by emilhe** [comment]
No, sorry, i have been caught up in with other projects :)

As i understand, the performance gain you have experienced is specific to (specific) plotly objects (e.g. graphs). And since a dash component can be _anything_, i am not a big fan of adopting the plotly-serialize-deserialize approach per default. In some some cases it will lower performance, and if the object(s) are not serializeable, it might event break existing apps. 

Personally, i don't really see the big different between passing a (pre defined) function handle or a string (assuming both are mentioned in the docstring), but i don't see any issue in the string approach either. 

Haha, i like the argument on why not to allow generic preprocessing functions :P. But in general i really like it when other people make their libraries flexible (like the option to add generic pre processing) so that i can adopt to my use case immediately and get going rather than having to file an issue and wait for a patch.

So, at the moment, i am leaning towards,

* Keeping the default behavior for "True" (which will also ensure backwards compability)
* Adding string options for the predefined functions, which will just be the string "plotly_jsonify" for now
* Keeping the option to pass in function handles (if people don't understand it, they can just avoid using it)

**Comment by sdementen** [comment]
@emilhe shouldn't this issue be closed?
You did implement the feature right?

**Comment by emilhe** [comment]
Yes! I finally got it merged into master yesterday. It's thus available in the 0.0.32 release, and a small example can be found here,

https://github.com/thedirtyfew/dash-extensions/blob/master/examples/callback_prejson.py

I am still amazed of the performance difference. On my current laptop, the default memoize takes 1.5s to update, but only 300 ms with the prejson option. That's 5X faster (:

**Comment by sdementen** [comment]
You also may look at https://community.plotly.com/t/speeding-up-plotting-large-timeseries-x5/47638
I have added some extra speedups vs what I have posted that accelerates even more. I will share them later
